astropy/units/quantity_helper/function_helpers.py

```diff
@@ -189,13 +189,15 @@ def invariant_x_helper(x, *args, **kwargs):
     return (x.view(np.ndarray),) + args, kwargs, x.unit, None
 
 
-# Note that ones_like does *not* work by default (unlike zeros_like) since if
-# one creates an empty array with a unit, one cannot just fill it with unity.
-# Indeed, in this respect, it is a bit of an odd function for Quantity. On the
-# other hand, it matches the idea that a unit is the same as the quantity with
-# that unit and value of 1. Also, it used to work without __array_function__.
-@function_helper
-def ones_like(a, *args, **kwargs):
+# Note that ones_like does *not* work by default since if one creates an empty
+# array with a unit, one cannot just fill it with unity.  Indeed, in this
+# respect, it is a bit of an odd function for Quantity. On the other hand, it
+# matches the idea that a unit is the same as the quantity with that unit and
+# value of 1. Also, it used to work without __array_function__.
+# zeros_like does work by default for regular quantities, but not for structured
+# dtype, so we include it here too.
```

Comment:
I now included a more detailed description (I had to remove it to find out what the problem was, so clearly good to actually describe it better!):
```
# zeros_like does work by default for regular quantities, because numpy first
# creates an empty array with the unit and then fills it with 0 (which can have
# any unit), but for structured dtype this fails (0 cannot have an arbitrary
# structured unit), so we include it here too.
```